src/editors/documentController.ts

```diff
@@ -0,0 +1,262 @@
+  }
+
+  async replaceDocument(
+    documentId: EJSON.SerializableTypes,
+    namespace: string,
+    dataservice: DataServiceType,
+    editor: vscode.TextEditor
+  ): Promise<boolean> {
+    const activeEditor = vscode.window.activeTextEditor;
+    let newDocument: EJSON.SerializableTypes = {};
+
+    try {
+      newDocument = EJSON.parse(editor.document.getText());
+    } catch (error) {
+      return this._saveDocumentFailed(error.message);
+    }
+
+    this._statusView.showMessage('Saving document...');
+
+    dataservice.findOneAndReplace(
+      namespace,
+      {
+        _id: documentId
+      },
+      newDocument,
+      {
+        returnOriginal: false
+      },
+      (error, data) => {
+        this._statusView.hideMessage();
+
+        if (error) {
+          return this._saveDocumentFailed(
+            `Unable to save document: ${error.message}`
+          );
+        }
+
+        // Send metrics to Segment.
+        this._telemetryController.trackDocumentUpdated('treeview', true);
+
+        // Save document changes to active editor.
+        activeEditor?.document.save();
+
+        // Update parent list of documents to reflect the changes that were made.
+        vscode.commands.executeCommand(
+          EXTENSION_COMMANDS.MDB_REFRESH_PLAYGROUND_RESULT_CONTENT,
+          data
+        );
+
+        vscode.window.showInformationMessage(
+          `The document was saved successfully to '${namespace}'`
+        );
+
+        return true;
+      }
+    );
+
+    return true;
+  }
+
+  async openEditableDocument(
+    documentId: EJSON.SerializableTypes,
+    namespace: string
+  ): Promise<boolean> {
+    const documentLocation = `${DOCUMENT_LOCATION_URI_IDENTIFIER}=mongodb`;
+    const connectionId = this._connectionController.getActiveConnectionId();
+    const connectionIdUriQuery = `${CONNECTION_ID_URI_IDENTIFIER}=${connectionId}`;
+    const documentIdReference = this._documentIdStore.add(documentId);
+    const documentIdUriQuery = `${DOCUMENT_ID_URI_IDENTIFIER}=${documentIdReference}`;
+    const namespaceUriQuery = `${NAMESPACE_URI_IDENTIFIER}=${namespace}`;
+    const localDocPath: string = path.join(
+      os.tmpdir(),
+      'vscode-opened-documents',
+      `${documentIdReference}.json`
+    );
+    const document = await this.provideDocumentContent(
+      namespace,
+      documentId,
+      connectionId
+    );
+
+    await fse.ensureFile(localDocPath);
+    await fse.writeJson(localDocPath, document, {
+      spaces: 2,
+      EOL: os.EOL
+    });
+
+    const uri: vscode.Uri = vscode.Uri.file(localDocPath).with({
+      query: `?${documentLocation}&${namespaceUriQuery}&${connectionIdUriQuery}&${documentIdUriQuery}`
+    });
+
+    return new Promise(async (resolve, reject) => {
+      vscode.workspace.openTextDocument(uri).then((doc) => {
+        vscode.window
+          .showTextDocument(doc, { preview: false, preserveFocus: true })
+          .then(() => resolve(true), reject);
+      }, reject);
```

Comment:
How do you feel about making this use async/await?
```ts
try {
  const doc = await vscode.workspace.openTextDocument(uri);
  await  vscode.window.showTextDocument(doc, { preview: false, preserveFocus: true });
} catch (err: Error) {
   throw err;
}
return true;
```

Reply:
It was written like this initially but sometimes caused an uncaught promise rejection error that was coming from somewhere. I also would like to see it async/await, but let's come back to it later in a separate PR for better code readability and additional testing.